<h1><font color="#113D68" size=6>Deep Learning en Python con PyTorch</font></h1>

<h1><font color="#113D68" size=5>Parte 4. Mejorar la Generalización en modelos</font></h1>

<h1><font color="#113D68" size=4>3. Funciones de pérdida</font></h1>

<br><br>
<div style="text-align: right">
<font color="#113D68" size=3>Manuel Castillo Cara</font><br>

</div>

<div class="alert alert-block alert-info">
    
<i class="fa fa-info-circle" aria-hidden="true"></i>
More information about [Manuel Castillo-Cara](https://www.manuelcastillo.eu/)

<div class="alert alert-block alert-info">

<i class="fa fa-info-circle" aria-hidden="true"></i>
Puedes ver más cursos de Inteligencia Artificial, Machine Learning y Deep Learning en mi [página web](https://www.manuelcastillo.eu/udemy/)


---

<a id="indice"></a>
<h2><font color="#004D7F" size=5>Índice</font></h2>

* [0. Contexto](#section0)
* [1. ¿Qué son las funciones de pérdida?](#section1)
* [2. Función de pérdida para regresión](#section2)
* [3. Función de pérdida para clasificación](#section3)
    * [3.1. Categorical cross-entropy](#section31)
    * [3.2. Clasificación multiclase con `NLLLoss`](#section32)
    * [3.3. Binary cross-entropy](#section33)
* [4. Funciones de Pérdida personalizada](#section4)

---
<a id="section0"></a>
# <font color="#004D7F" size=6> 0. Contexto</font>

La métrica de pérdida es muy importante para las redes neuronales. Dado que todos los modelos de aprendizaje automático son, en esencia, un problema de optimización, la pérdida es la función objetivo que se debe minimizar. En las redes neuronales, la optimización se realiza con el descenso de gradiente y la retropropagación. Pero ¿qué son exactamente las funciones de pérdida y cómo afectan a las redes neuronales?

En este capítulo, aprenderás qué son las funciones de pérdida, explorarás algunas de las más utilizadas y cómo puedes aplicarlas en tus redes neuronales. Después de leer este capítulo, comprenderás:

- Qué son las funciones de pérdida y su papel en el entrenamiento de modelos de redes neuronales.  
- Funciones de pérdida comunes para problemas de regresión y clasificación.  
- Cómo usar funciones de pérdida en tu modelo PyTorch.  

<div class="alert alert-block alert-info">
    
<i class="fa fa-info-circle" aria-hidden="true"></i>
[PyTorch tutorials.](https://pytorch.org/tutorials/)

<div class="alert alert-block alert-info">
    
<i class="fa fa-info-circle" aria-hidden="true"></i>
PyTorch API [`torch.nn`.](https://pytorch.org/docs/stable/nn.html)

<div class="alert alert-block alert-info">
    
<i class="fa fa-info-circle" aria-hidden="true"></i>
PyTorch API [`torch.optim`.](https://pytorch.org/docs/stable/optim.html)

---
<div style="text-align: right"> <font size=5> <a href="#indice"><i class="fa fa-arrow-circle-up" aria-hidden="true" style="color:#004D7F"></i></a></font></div>

---

<a id="section1"></a>
# <font color="#004D7F" size=6>1. ¿Qué son las funciones de pérdida?</font>

- Funciones de pérdida en redes neuronales:
  - Ayudan a optimizar el rendimiento del modelo

  - Miden la penalización por predicciones incorrectas

  - Generalmente miden la desviación entre predicción y etiqueta verdadera

- Uso en el ciclo de entrenamiento:
  - Se diferencian respecto a los parámetros del modelo

  - Las derivadas se usan en retropropagación y descenso de gradiente

  - Objetivo: optimizar el modelo en el conjunto de entrenamiento

- Diferencia entre funciones de pérdida y métricas:
  - Funciones de pérdida: pueden no ser de interés directo o fácilmente explicables

  - Métricas (ej. accuracy): más comprensibles para humanos

  - Métricas no siempre son buenas funciones de pérdida (pueden no ser diferenciables)

- Tipos de funciones de pérdida a explorar:
  - Para problemas de regresión

  - Para problemas de clasificación

- Importancia:
  - Cruciales para el entrenamiento efectivo de redes neuronales
  
  - La elección adecuada impacta significativamente en el rendimiento del modelo

---
<div style="text-align: right"> <font size=5> <a href="#indice"><i class="fa fa-arrow-circle-up" aria-hidden="true" style="color:#004D7F"></i></a></font></div>

---

<a id="section2"></a>
# <font color="#004D7F" size=6>2. Funciones de Pérdida para Regresión</font>

- Problemas de regresión:
  - Objetivo: Predecir valor en rango continuo

  - Ideal: Predicción exacta, pero cercanía es suficiente

- Función de pérdida para regresión:
  - Mide cercanía entre predicción y valor real
  
  - Mayor pérdida para predicciones más lejanas

- Mean Absolute Error (MAE):
  - Fórmula: $\text{MAE} = \frac{1}{m} \sum_{i=1}^{m} | \hat{y}_i - y_i |$

  - Características:

    - Siempre no negativa

    - Cero solo con predicción perfecta
    
    - Intuitiva y útil como métrica

- Mean Squared Error (MSE):
  - Fórmula: $\text{MSE} = \frac{1}{m} \sum_{i=1}^{m} ( \hat{y}_i - y_i )^2$

  - Ventajas sobre MAE:

    - Diferenciable en x = 0

    - Tolera mejor errores pequeños

    - Amplifica errores grandes

- Comparación gráfica MAE vs MSE:
  - MAE: Gradiente constante (excepto en x = 0)

  - MSE: Gradiente mayor para errores grandes

- Implementación en PyTorch:
  - MAE: `nn.L1Loss()`
  
  - MSE: `nn.MSELoss()`

<div class="alert alert-block alert-info">
    
<i class="fa fa-info-circle" aria-hidden="true"></i>
Más información sobre el dataset [`L1Loss`](https://pytorch.org/docs/stable/generated/torch.nn.L1Loss.html)

<div class="alert alert-block alert-info">
    
<i class="fa fa-info-circle" aria-hidden="true"></i>
Más información sobre el dataset [`MSELoss`](https://pytorch.org/docs/stable/generated/torch.nn.MSELoss.html)

In [1]:
import torch
import torch.nn as nn

mae = nn.L1Loss()  # MAE
mse = nn.MSELoss()

predict = torch.tensor([0., 3.])
target = torch.tensor([1., 0.])

print("MAE: %.3f" % mae(predict, target))
print("MSE: %.3f" % mse(predict, target))

MAE: 2.000
MSE: 5.000


En este ejemplo, el MAE es 2.0 porque:

$$
\frac{1}{2} \left( |0 - 1| + |3 - 0| \right) = \frac{1}{2}(1 + 3) = 2
$$

Mientras que el MSE es 5.0 porque:

$$
\frac{1}{2} \left( (0 - 1)^2 + (3 - 0)^2 \right) = \frac{1}{2}(1 + 9) = 5
$$

Esto muestra que en el MSE, el segundo ejemplo (predicción de 3 con un valor objetivo de 0) contribuye con el 90% del error frente al 75% bajo el MAE.

A veces, también se utiliza el *Root Mean Squared Error* (RMSE) como métrica. 
- El RMSE toma la raíz cuadrada del MSE. Desde la perspectiva de una función de pérdida, el MSE y el RMSE son equivalentes. 

- Sin embargo, el RMSE está en la misma unidad que las predicciones, lo que facilita la interpretación. 

---
<div style="text-align: right"> <font size=5> <a href="#indice"><i class="fa fa-arrow-circle-up" aria-hidden="true" style="color:#004D7F"></i></a></font></div>

---

<a id="section3"></a>
# <font color="#004D7F" size=6>3. Funciones de Pérdida para Clasificación</font>

- Problemas de clasificación:
  - Objetivo: Predecir una categoría de un conjunto discreto de clases

- Características de las etiquetas de clase:
  - Codificadas con números

  - Los números no tienen significado semántico

  - Ejemplo: 0 para "gato", 1 para "perro", 2 para "caballo"

- Implicaciones:
  - La codificación numérica no debe afectar el rendimiento del modelo

  - No hay relación aritmética entre las clases (ej. "perro" no es mitad "gato" y mitad "caballo")

- Consideraciones para funciones de pérdida:
  - Deben tratar las clases de manera equitativa

  - No deben asumir relaciones ordinales entre las clases

- Diferencias con problemas de regresión:
  - En regresión, los valores tienen significado numérico

  - En clasificación, los valores son solo etiquetas sin orden inherente

- Importancia en el diseño del modelo:
  - La función de pérdida debe reflejar la naturaleza discreta de las clases

  - Debe penalizar igualmente las clasificaciones incorrectas, independientemente de la codificación numérica

<a id="section31"></a>
# <font color="#004D7F" size=5>3.1. Categorical cross-entropy</font>

- Salida del modelo en problemas de clasificación:
  - Vector de probabilidades para cada categoría

  - Formato esperado: logits o resultado de activación softmax

- Entropía cruzada:
  - Mide diferencia entre dos distribuciones de probabilidad

  - Fórmula general: $-\sum_i P(X = x_i) \log Q(X = x_i)$

    - _P_: probabilidad real

    - _Q_: probabilidad predicha

- Simplificación para aprendizaje automático:
  - $\text{Entropía cruzada categórica} = -\log p_{\text{target}}$

  - $p_{\text{target}}$: probabilidad predicha para la clase verdadera

- Características de la entropía cruzada:
  - Signo negativo compensa tendencia de log(x) a -∞

  - Pérdida aumenta cuando la probabilidad se acerca a 0

  - Penaliza fuertemente las malas predicciones

- Implementación en PyTorch:
  - Función: `nn.CrossEntropyLoss()`
  
  - Características:

    - Toma logits predichos y objetivos como parámetros

    - Aplica softmax internamente a los logits

    - No requiere softmax en la capa de salida del modelo

<div class="alert alert-block alert-info">
    
<i class="fa fa-info-circle" aria-hidden="true"></i>
Más información sobre el dataset [`CrossEntropyLoss`](https://pytorch.org/docs/stable/generated/torch.nn.CrossEntropyLoss.html)

In [2]:
ce = nn.CrossEntropyLoss()  # espera LOGITS (aplica softmax internamente)

logits = torch.tensor([[-1.90, -0.29, -2.30], [-0.29, -1.90, -2.30]])
target = torch.tensor([[0., 1., 0.], [1., 0., 0.]])  # one-hot (o probabilidades)
print("Cross entropy: %.3f" % ce(logits, target))

Cross entropy: 0.288




Este valor de 0.288 proviene de $-\log(0.75)$, ya que, al aplicar la función softmax a los logits, obtenemos probabilidades como:

```python
    probs = torch.tensor([[0.15, 0.75, 0.1], [0.75, 0.15, 0.1]])
```

Otra forma de calcular la entropía cruzada es usando etiquetas enteras en lugar de one-hot encoding. 

In [3]:
logits = torch.tensor([[-1.90, -0.29, -2.30], [-0.29, -1.90, -2.30]])
indices = torch.tensor([1, 0])  # índice de la clase correcta
print("Cross entropy: %.3f" % ce(logits, indices))

Cross entropy: 0.288


Esto también devuelve la misma entropía cruzada de 0.288. Nota lo que devuelve:

In [4]:
target = torch.tensor([[0., 1., 0.], [1., 0., 0.]])
indices = torch.argmax(target, dim=1)
print(indices)

tensor([1, 0])


<a id="section32"></a>
# <font color="#004D7F" size=5>3.2. Clasificación multiclase con `NLLLoss`</font>

- Interpretación de tensor objetivo en PyTorch:
  - También conocido como función de entropía cruzada dispersa en otras bibliotecas

  - No espera un vector one-hot como entrada

- Uso de `nn.LogSoftmax()` en PyTorch:
  - Función de activación que realiza softmax y luego toma el logaritmo de cada elemento

  - Se aplica en la capa de salida del modelo

- Combinación de `LogSoftmax` y `NLLLoss`:
  - Si se usa `LogSoftmax` en la capa de salida, se debe usar `nn.NLLLoss()` (pérdida de verosimilitud logarítmica negativa) como función de pérdida

  - Matemáticamente, esta combinación es equivalente a la pérdida de entropía cruzada

<div class="alert alert-block alert-info">
    
<i class="fa fa-info-circle" aria-hidden="true"></i>
Más información sobre el dataset [`NLLLoss`](https://pytorch.org/docs/stable/generated/torch.nn.NLLLoss.html)

In [5]:
ce = nn.NLLLoss()

logsoftmax = nn.LogSoftmax(dim=1)

logits = torch.tensor([[-1.90, -0.29, -2.30], [-0.29, -1.90, -2.30]])
pred = logsoftmax(logits)
indices = torch.tensor([1, 0])
print("Cross entropy: %.3f" % ce(pred, indices))

Cross entropy: 0.288


<a id="section33"></a>
# <font color="#004D7F" size=5>3.3. Binary cross-entropy</font>

- Clasificación binaria:
  - Problema de clasificación con solo 2 clases

  - Equivalente a un modelo de regresión logística

- Características especiales:
  - Una sola salida en lugar de un vector de dos valores

  - Probabilidad de "clase positiva" (x) implica probabilidad de "clase negativa" (1-x)

- PyTorch: `nn.BCELoss()` para entropía cruzada binaria

- Consideraciones:
  - Asegurarse de que las predicciones estén en el rango [0, 1]

  - Las etiquetas deben ser 0 o 1 para clasificación binaria

- Ventajas de usar `BCELoss`:
  - Más eficiente computacionalmente para problemas binarios

  - Interpretación directa de la salida como probabilidad

<div class="alert alert-block alert-info">
    
<i class="fa fa-info-circle" aria-hidden="true"></i>
Más información sobre el dataset [`BCELoss`](https://pytorch.org/docs/stable/generated/torch.nn.BCELoss.html)

In [6]:
bce = nn.BCELoss()

pred = torch.tensor([0.75, 0.25])
target = torch.tensor([1., 0.])
print("Binary cross entropy: %.3f" % bce(pred, target))

Binary cross entropy: 0.288


Esto se debe a que:

$$
\frac{1}{2} \left( -\log(0.75) + -\log(1 - 0.25) \right) = -\log(0.75) = 0.288
$$

En PyTorch, el valor 1 en la etiqueta de destino representa la "clase positiva" y el 0 representa la "clase negativa". No se permiten otros valores en el tensor de destino.

---
<div style="text-align: right"> <font size=5> <a href="#indice"><i class="fa fa-arrow-circle-up" aria-hidden="true" style="color:#004D7F"></i></a></font></div>

---

<a id="section4"></a>
# <font color="#004D7F" size=6>4. Funciones de Pérdida personalizada</font>

- Flexibilidad en la creación de funciones de pérdida en PyTorch:
  - Las métricas de pérdida se calculan con objetos de `torch.nn`

  - El resultado es un tensor de PyTorch, permitiendo diferenciación y retropropagación

  - Posibilidad de crear funciones de pérdida personalizadas

- Ejemplo de métrica no incluida en PyTorch: ***MAPE (Mean Absolute Percentage Error)***
  - Similar al *MAE (Mean Absolute Error)*

  - Fórmula: $\text{MAPE} = \frac{1}{m} \sum_{i=1}^{m} \left| \frac{\hat{y_i} - y_i}{y_i} \right|$

- Ventajas del MAPE:
  - Útil en ciertos contextos, como predicción de precios de viviendas

  - Evalúa accuracy basada en diferencia porcentual en lugar de absoluta

- Consideraciones al crear funciones de pérdida personalizadas:
  - Asegurar que la función sea diferenciable

  - Manejar casos especiales (como división por cero en MAPE)

  - Verificar que la función escale adecuadamente con el tamaño del batch

In [7]:
import copy
import numpy as np
import torch
import torch.nn as nn
import torch.optim as optim
from sklearn.model_selection import train_test_split
from sklearn.datasets import fetch_california_housing
from sklearn.preprocessing import StandardScaler

data = fetch_california_housing()
X, y = data.data, data.target

X_train_raw, X_test_raw, y_train, y_test = train_test_split(X, y, train_size=0.7, shuffle=True)

# estandarizar (ajustando el escalador SOLO con train)
scaler = StandardScaler()
scaler.fit(X_train_raw)
X_train = scaler.transform(X_train_raw)
X_test = scaler.transform(X_test_raw)

X_train = torch.tensor(X_train, dtype=torch.float32)
y_train = torch.tensor(y_train, dtype=torch.float32).reshape(-1, 1)
X_test = torch.tensor(X_test, dtype=torch.float32)
y_test = torch.tensor(y_test, dtype=torch.float32).reshape(-1, 1)

model = nn.Sequential(
    nn.Linear(8, 24),
    nn.ReLU(),
    nn.Linear(24, 12),
    nn.ReLU(),
    nn.Linear(12, 6),
    nn.ReLU(),
    nn.Linear(6, 1)
)

########### Agreamos la función MAPE #############
# una función de pérdida es solo una función que devuelve un tensor: autograd hace el resto
def loss_fn(output, target):
    return torch.mean(torch.abs((target - output) / target))  # MAPE

optimizer = optim.Adam(model.parameters(), lr=0.0001)

n_epochs = 100
batch_size = 10
batch_start = torch.arange(0, len(X_train), batch_size)

best_mape = np.inf
best_weights = None

for epoch in range(n_epochs):
    model.train()
    for start in batch_start:
        X_batch = X_train[start:start+batch_size]
        y_batch = y_train[start:start+batch_size]

        y_pred = model(X_batch)
        loss = loss_fn(y_pred, y_batch)

        optimizer.zero_grad()
        loss.backward()
        optimizer.step()
    model.eval()
    with torch.no_grad():
        y_pred = model(X_test)
        mape = float(loss_fn(y_pred, y_test))
    if mape < best_mape:
        best_mape = mape
        best_weights = copy.deepcopy(model.state_dict())

model.load_state_dict(best_weights)
print("MAPE: %.2f%%" % (best_mape * 100))

model.eval()
with torch.no_grad():
    for i in range(5):
        X_sample = X_test_raw[i: i+1]
        X_sample = scaler.transform(X_sample)
        X_sample = torch.tensor(X_sample, dtype=torch.float32)
        y_pred = model(X_sample)
        print(f"{X_test_raw[i]} --> {y_pred[0].numpy()} (esperado {y_test[i].numpy()})")

MAPE: 19.13%
[   2.375        37.            4.1440678     1.24858757  651.
    1.83898305   32.79       -117.25      ] --> [1.8299851] (esperado [3.404])
[ 3.44500000e+00  8.00000000e+00  4.78412391e+00  1.13552759e+00
  3.01200000e+03  2.91577928e+00  3.42700000e+01 -1.19200000e+02] --> [1.7534835] (esperado [2.034])
[   1.5809       34.            3.51807229    1.06024096  625.
    3.76506024   33.68       -116.21      ] --> [0.5112806] (esperado [1.])
[ 5.45270000e+00  3.30000000e+01  6.42610837e+00  1.06157635e+00
  1.20800000e+03  2.97536946e+00  3.42100000e+01 -1.18610000e+02] --> [2.8062775] (esperado [2.271])
[ 4.53820000e+00  3.90000000e+01  4.97256098e+00  9.81707317e-01
  1.03400000e+03  3.15243902e+00  3.39200000e+01 -1.18080000e+02] --> [1.9247109] (esperado [1.657])


<div style="text-align: right"> <font size=5> <a href="#indice"><i class="fa fa-arrow-circle-up" aria-hidden="true" style="color:#004D7F"></i></a></font></div>

---

<div style="text-align: right"> <font size=6><i class="fa fa-coffee" aria-hidden="true" style="color:#004D7F"></i> </font></div>